src/Microsoft.TestPlatform.CrossPlatEngine/Adapter/TestExecutionRecorder.cs

```diff
@@ -93,7 +133,41 @@ public void RecordEnd(TestCase testCase, TestOutcome outcome)
 
             if (this.testCaseEventsHandler != null)
             {
-                this.testCaseEventsHandler.SendTestCaseEnd(testCase, outcome);
+                var isTestCaseEndAlreadySent = false;
+
+                lock (this.testCaseEndStatusSyncObject)
+                {
+                    isTestCaseEndAlreadySent = this.testCaseEndStatusMap.Contains(testCase.Id);
+                    if (!isTestCaseEndAlreadySent)
+                    {
+                        this.testCaseEndStatusMap.Add(testCase.Id);
+                    }
+
+                    // Do not support multiple - testcasends for a single test case start
+                    // TestCaseEnd must always be preceded by testcasestart for a given test case id
+                    if (!isTestCaseEndAlreadySent)
```

Comment:
General verification done using xunit.integration project. 
Also verified Theory, DataRow and Fixtures.